src/Analysis/Engine/Test/CompletionTests.cs

```diff
@@ -0,0 +1,1056 @@
+                await AssertCompletion(s, modP2, new[] { "y", "z" }, new[] { "x" });
+                await AssertCompletion(s, modP3, new[] { "y", "z" }, new[] { "x" });
+            }
+        }
+
+        private static async Task AssertCompletion(Server s, Uri uri, IReadOnlyCollection<string> contains, IReadOnlyCollection<string> excludes, Position? position = null, CompletionContext? context = null, Func<CompletionItem, string> cmpKey = null, string expr = null, Range? applicableSpan = null) {
+            await s.WaitForCompleteAnalysisAsync(CancellationToken.None);
+            var res = await s.Completion(new CompletionParams {
+                textDocument = new TextDocumentIdentifier { uri = uri },
+                position = position ?? new Position(),
+                context = context,
+                _expr = expr
+            }, CancellationToken.None);
+            DumpDetails(res);
+
+            cmpKey = cmpKey ?? (c => c.insertText);
+            var items = res.items?.Select(cmpKey).ToList() ?? new List<string>();
+
+            if (contains != null && contains.Any()) {
+                items.Should().Contain(contains);
+            }
+
+            if (excludes != null && excludes.Any()) {
+                items.Should().NotContain(excludes);
+            }
+
+            if (applicableSpan.HasValue) {
+                res._applicableSpan.Should().Be(applicableSpan);
+            }
+        }
+
+        private static async Task AssertAnyCompletion(Server s, TextDocumentIdentifier document, Position position) {
+            await s.WaitForCompleteAnalysisAsync(CancellationToken.None).ConfigureAwait(false);
+            var res = await s.Completion(new CompletionParams { textDocument = document, position = position }, CancellationToken.None);
+            DumpDetails(res);
+            if (res.items == null || !res.items.Any()) {
+                Assert.Fail("Completions were not returned");
+            }
+        }
+
+        private static async Task AssertNoCompletion(Server s, TextDocumentIdentifier document, Position position) {
+            await s.WaitForCompleteAnalysisAsync(CancellationToken.None).ConfigureAwait(false);
+            var res = await s.Completion(new CompletionParams { textDocument = document, position = position }, CancellationToken.None);
+            DumpDetails(res);
+            if (res.items != null && res.items.Any()) {
+                var msg = string.Join(", ", res.items.Select(c => c.label).Ordered());
+                Assert.Fail("Completions were returned: " + msg);
+            }
+        }
+
+        private static void DumpDetails(CompletionList completions) {
+            var span = ((SourceSpan?)completions._applicableSpan) ?? SourceSpan.None;
+            Debug.WriteLine($"Completed {completions._expr ?? "(null)"} at {span}");
+        }
+
+        class TestCompletionHookProvider : ILanguageServerExtensionProvider {
```

Comment:
Few reasons why they are here:
- Test uses specifically `ICompletionExtension` since extensions can implement stuff selectively
- It tests only completions
- It relies on `AssertCompletion`